### PR TITLE
Feat: Auto-completion of @-attributes for the language server

### PR DESCRIPTION
**Pull request by MikaelMayer**
### Description
This PR adds the ability for detecting that the user presses @ when after a space, which is the signature of an attribute. It then provides auto-complete of the attribute name.

### How has this been tested?
Tests added in DafnyLanguageServer.Tests

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs [hunk 1]
```diff
@@ -122,6 +199,15 @@ private static bool IsConstructor(ILegacySymbol symbol) {
         return symbol is MethodSymbol method && method.Name == "_ctor";
       }
 
+      private CompletionItem CreateCompletionItem(string attributeName) {
+        return new CompletionItem {
+          Label = attributeName,
+          Kind = CompletionItemKind.Constructor,
+          InsertText = attributeName, // TODO: Parentheses
+          Detail = "" // TODO: Details of attribute name
```
Comment (robin-aws) [inline Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs:207]: These look more like unfinished PR todos rather than planned improvements for the next iteration

File: Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs [hunk 2]
```diff
@@ -86,11 +93,81 @@ public CompletionList Process() {
         return new CompletionList();
       }
 
+      private string GetLastTwoCharactersBeforePositionIncluded(TextReader fileContent, DafnyPosition position) {
```
Comment (robin-aws) [inline Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs:96]: I would think this would be a lot simpler if you located the index of the current position and then grabbed the previous two characters. It doesn't seem like processing anything line-by-line buys you anything - "\n@" will still be processed correctly since `\n` is whitespace.

File: Source/DafnyLanguageServer.Test/Completion/AtCompletionTest.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.IntegrationTest.Util;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Dafny.LanguageServer.IntegrationTest.Completion {
```
Comment (robin-aws) [inline Source/DafnyLanguageServer.Test/Completion/AtCompletionTest.cs:11]: You should have at least one test case where `@` is the very first character of the file (i.e. not even any comment first), since that's a logic edge case.

**Review by robin-aws — CHANGES_REQUESTED** [review]

**Review by keyboardDrummer — CHANGES_REQUESTED** [review]

File: Source/DafnyLanguageServer/Workspace/ProjectManager.cs [hunk 4]
```diff
@@ -126,6 +127,10 @@ public ProjectManager(
 
   private const int MaxRememberedChanges = 100;
 
+  public TextReader ReadFile(Uri fileToRead) {
```
Comment (keyboardDrummer) [inline Source/DafnyLanguageServer/Workspace/ProjectManager.cs:130]: This method here makes no sense to me. Why not use a filesystem directly?

File: Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs [hunk 5]
```diff
@@ -86,11 +93,81 @@ public CompletionList Process() {
...
         var node = state.Program.FindNode<INode>(request.TextDocument.Uri.ToUri(), request.Position.ToDafnyPosition());
         return node?.RangeToken.EndToken.val == ".";
       }
 
+      private bool IsAtAttribute() {
+        var position = request.Position.ToDafnyPosition();
+        if (projectManager == null) {
+          return false;
+        }
+        var fileContent = projectManager.ReadFile(request.TextDocument.Uri.ToUri());
```
[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs]: I'll get back to this soon, but meantime, do you have a hint about where the already processed file contents is? I thought ReadFile would just give me the current string

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs]: I've pushed a commit that addresses this comment.

File: Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs [hunk 6]
```diff
@@ -86,11 +93,81 @@ public CompletionList Process() {
...
+        var lastTwoChars = GetLastTwoCharactersBeforePositionIncluded(fileContent, position);
+        var isAtAttribute =
+          lastTwoChars == "@" ||
+          lastTwoChars.Length >= 2 && lastTwoChars[1] == '@' && char.IsWhiteSpace(lastTwoChars[0]);
+        return isAtAttribute;
+      }
+
+      private CompletionList CreateAtAttributeCompletionList() {
+        var completionItems =
+          Attributes.BuiltinAtAttributes.Select(b =>
```
[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs:126]: I'm only providing instant code completion upon writing the @, not afterwards. The IDE will then filter the list according to what is being typed.

File: Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MikaelMayer) [reply Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs]: Do you know why your commit breaks some tests? For example:

Could not close file file:///tmp/z3sq5jsc.kyh/DocumentStaysUnloadedWhenClosed.dfy because it was not open.

File: Source/DafnyLanguageServer/Workspace/ProjectManager.cs (hunk 4 above)
Comment (MikaelMayer) [reply Source/DafnyLanguageServer/Workspace/ProjectManager.cs]: Your update for reading the last two characters forgot the test case when the position is given in line zero but the column is the number of characters. That fails the test slowlytypefile

**Review by keyboardDrummer — APPROVED** [review]

File: Source/DafnyLanguageServer/Workspace/ProjectManager.cs (hunk 4 above)
Comment (keyboardDrummer) [reply Source/DafnyLanguageServer/Workspace/ProjectManager.cs]: I added a fix for `slowlytypefile`, although I don't recognize the bug you describe.

**Review by robin-aws — DISMISSED** [review]

**Review by keyboardDrummer — DISMISSED** [review]

**Review by keyboardDrummer — APPROVED** [review]
